providers/flagd/src/main/java/dev/openfeature/contrib/providers/flagd/FlagdProvider.java

```diff
@@ -527,6 +535,30 @@ private static FieldDescriptor getFieldDescriptor(Message message, String name)
         return message.getDescriptorForType().findFieldByName(name);
     }
 
+    private static ImmutableMetadata metadataFromResponse(Message response) {
+        final Object metadata = response.getField(getFieldDescriptor(response, METADATA_FIELD));
+
+        if (!(metadata instanceof Struct)) {
+            return ImmutableMetadata.builder().build();
+        }
+
+        final Struct struct = (Struct) metadata;
+
+        ImmutableMetadata.ImmutableMetadataBuilder builder = ImmutableMetadata.builder();
+
+        for (Map.Entry<String, com.google.protobuf.Value> entry : struct.getFieldsMap().entrySet()) {
+            if (entry.getValue().hasStringValue()) {
+                builder.addString(entry.getKey(), entry.getValue().getStringValue());
+            } else if (entry.getValue().hasBoolValue()) {
+                builder.addBoolean(entry.getKey(), entry.getValue().getBoolValue());
+            } else if (entry.getValue().hasNumberValue()) {
+                builder.addDouble(entry.getKey(), entry.getValue().getNumberValue());
+            }
+        }
```

Comment:
I like streams but that's personal preference (and probably slower?)

Reply:
IMO streaming here is not needed as we simply wants to iterate over the entries. However, if there's some filtering, then streams would help.

I don't think there's any performance impact for smaller collection from either approach